frappe/core/api/user_invitation.py

```diff
@@ -0,0 +1,110 @@
+import frappe
+
+USER_INVITATION_DOCTYPE = "User Invitation"
+FRAMEWORK_APP_NAME = " "
+
+
+@frappe.whitelist()
+def invite_by_email(
+	emails: str, role: str, redirect_to_path: str, app_name: str = FRAMEWORK_APP_NAME
+) -> dict[str, list[str]]:
+	frappe.only_for(["System Manager", "User Invitation Manager"])
+	is_app_framework = app_name == FRAMEWORK_APP_NAME
+	allowed_roles: list[str] = []
+	if not is_app_framework:
+		# `role` validation
+		invalid_app_msg = frappe._("Invalid app")
+		try:
+			user_invitation_hook = frappe.get_hooks("user_invitation", app_name=app_name)
+			allowed_invite_roles_key = "allowed_invite_roles"
+			if (
+				not isinstance(user_invitation_hook, dict)
+				or not isinstance(user_invitation_hook.get(allowed_invite_roles_key), list)
+				or not all(isinstance(r, str) for r in user_invitation_hook.get(allowed_invite_roles_key))
+			):
+				frappe.throw(invalid_app_msg)
+			allowed_roles = user_invitation_hook.get(allowed_invite_roles_key)
+			if not allowed_roles:
+				frappe.throw(invalid_app_msg)
+		except ImportError:
+			frappe.throw(invalid_app_msg)
+		if role not in allowed_roles:
+			frappe.throw(frappe._("Invalid role"))
+	frappe.utils.validate_email_address(emails, throw=True)
+	email_list = frappe.utils.split_emails(emails)
+	if not email_list:
+		frappe.throw(frappe._("No email addresses found"))
+	existing_user_emails = frappe.db.get_all("User", filters={"email": ["in", email_list]}, pluck="email")
+	existing_invited_emails_filters = {"email": ["in", email_list], "status": "Pending"}
+	if not is_app_framework:
+		existing_invited_emails_filters["role"] = ["in", allowed_roles]
+	existing_invited_emails = frappe.db.get_all(
+		USER_INVITATION_DOCTYPE, filters=existing_invited_emails_filters, pluck="email"
+	)
+	to_invite = list(set(email_list) - set(existing_user_emails) - set(existing_invited_emails))
+	for email in to_invite:
+		frappe.get_doc(
+			doctype=USER_INVITATION_DOCTYPE,
+			email=email,
+			role=role,
+			app_name=app_name,
+			redirect_to_path=redirect_to_path,
+		).insert()
+	return {
+		"existing_user_emails": existing_user_emails,
+		"existing_invited_emails": existing_invited_emails,
+		"invited_emails": to_invite,
+	}
+
+
+@frappe.whitelist(allow_guest=True)
+def accept_invitation(key: str) -> None:
+	result = frappe.db.get_all(USER_INVITATION_DOCTYPE, filters={"key": key}, pluck="name")
+	invalid_or_expired_key_msg = frappe._("Invalid or expired key")
+	if not result:
+		frappe.throw(invalid_or_expired_key_msg)
+	invitation = frappe.get_doc(USER_INVITATION_DOCTYPE, result[0])
```

Comment:
It's okay to not use variable name for doctypes. Keep it consistent with rest of codebase where doctype name is repeated wherever it's used. Better readability IMO.